src/RenderCache/Entity/CacheFragmentController.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\restful\RenderCache\Entity\CacheFragmentController.
+ */
+
+namespace Drupal\restful\RenderCache\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+
+class CacheFragmentController extends \EntityAPIController {
+
+  /**
+   * Creates all the caches tags from the tag collection.
+   *
+   * @param ArrayCollection $cache_fragments
+   *   The collection of tags.
+   *
+   * @return CacheFragment[]
+   *   An array of fragments.
+   */
+  public function createCacheFragments(ArrayCollection $cache_fragments) {
+    $hash = $this->generateCacheHash($cache_fragments);
+    if ($fragments = $this->existingFragments($hash)) {
+      return $fragments;
+    }
+    foreach ($cache_fragments as $tag_type => $tag_value) {
+      $cache_fragment = new CacheFragment(array(
+        'value' => $tag_value,
+        'type' => $tag_type,
+        'hash' => $hash,
+      ), 'cache_fragment');
+      try {
+        if ($id = $this->save($cache_fragment)) {
+          $fragments[] = $cache_fragment;
+        }
+      }
+      catch (\Exception $e) {
+        // Log the exception. It's probably a duplicate fragment.
+        watchdog_exception('restful', $e);
+      }
+    }
+    return $fragments;
+  }
+
+  /**
+   * Gets the existing fragments for a given hash.
+   *
+   * @param string $hash
+   *   The hash.
+   *
+   * @return CacheFragment[]
+   *   An array of fragments.
+   */
+  protected function existingFragments($hash) {
+    $query = new \EntityFieldQuery();
+    $results = $query
+      ->entityCondition('entity_type', 'cache_fragment')
+      ->propertyCondition('hash', $hash)
+      ->execute();
+    return empty($results['cache_fragment']) ? array() : $this->load(array_keys($results['cache_fragment']));
+  }
+
+  /**
+   * Generated the cache hash based on the cache fragments collection.
+   *
+   * @param ArrayCollection $cache_fragments
+   *   The collection of tags.
+   *
+   * @return string
+   *   The generated hash.
+   */
+  public function generateCacheHash(ArrayCollection $cache_fragments) {
+    return substr(sha1(serialize($cache_fragments->toArray())), 0, 40);
+  }
+
+  /**
+   * Gets the hashes for an EFQ.
+   *
+   * @param \EntityFieldQuery $query
+   *   The EFQ.
+   *
+   * @return string[]
+   *   The hashes that meet the conditions.
+   */
+  public static function lookUpHashes(\EntityFieldQuery $query) {
+    $results = $query->execute();
```

Comment:
This is (supposedly) only for developer experience. So they can tweak the query by overriding the controller. Not sure if that is ever gonna happen.